Curso Formação Cientista de Dados
===

# O Curso

Curso de Data Science você vai conhecer e aprender a aplicar todos os principais conceitos e técnicas para se qualificar e atuar como um Cientista de Dados.

Link do Curso:
https://www.udemy.com/cientista-de-dados/learn/lecture/11820720#overview

# 01. Estatística

Tipos:
- Descritiva
- Probabilidae
- Inferência


## Observação e Experimento

Observação: elementos não são afetados (pequisa)

Experimento: Condições para avaliar o resultado

## Variáveis

Quantitivas: numéricas:
- Contínuas
- Discretas

Qualitativas: categóricas
- Nominais
- Ordinais


## Amostra

Parte de uma população, selecionada usando alguma tecnica que de changes iguais a todos os elementos da população de serem selecionados


Conceitos
- População: alvo do estudo
- Amostra: subjconjunto da população
  - caro ou impossível inferir sobre toda a população
- Censo: pesquisa com toda a população


### É possível inferir sobre uma amostra

Correta:
- Representa a mesmas características da população

Enviesada
- Não representa a população

### Custo da Amostra

Margem de erro e nível de confiança 

Variação: amostras diferentes podem apresentar resultados dieferentes 

Podemos medir a variação esperada

## 01.1 Amostra

Separação da amostra para teste/treino/validação

Experimentos diferentes

### Tipos de Amostra

Aleatória Simples
- determinado número de elementos é retirado de forma aleatória
- mesmas chances para todos os elementos

Com Reposição:
- Após selecionado, o elemento pode voltar à população
- Teste anti-doping

Sem resposição:
- Após reposição, não volta
- Pesquisa de opinião

### Amostra Estretificada

População dvididas nos chamados estratos:
- número representativo de cada grupo na população

### Amostra Sistemática

Escolhe um número aleatório, a cada N elementos um novo será selecionado
- Não se sabe o número total da população

### Amostragem por Unidade Monetária

Duas colunas, uma de ordenação e outra numérica. Os registros são selecionados para a amostra toda vez que o valor for atingido.

Clientes e débitos


## Dados

Dados localizados na pasta:
`D:\_SIA_\estudo\udemy\cientista-de-dados\02-estatistica-01`


Arquivos para upload:
- 08.Dados.zip
- Disponível também através do link https://github.com/ronivaldo/udemy-curso-cientista_dados/raw/master/data/08.Dados.zip

Conteúdo do arquivo:
- infert.csv
- iris.csv



In [1]:
import os  

# verifica se upload ja foi realizado
if not os.path.isfile('./Dados.zip'):
  # obtem o arquivo do Drive
  !wget https://github.com/ronivaldo/udemy-curso-cientista_dados/raw/master/data/08.Dados.zip

# verifica se upload ja foi realizado
if not os.path.isfile('./infert.csv'):
  # descompacta o arquivo Dados.zip
  !unzip 08.Dados.zip

--2019-04-01 23:34:25--  https://github.com/ronivaldo/udemy-curso-cientista_dados/raw/master/data/08.Dados.zip
Resolving github.com (github.com)... 192.30.253.113, 192.30.253.112
Connecting to github.com (github.com)|192.30.253.113|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://raw.githubusercontent.com/ronivaldo/udemy-curso-cientista_dados/master/data/08.Dados.zip [following]
--2019-04-01 23:34:25--  https://raw.githubusercontent.com/ronivaldo/udemy-curso-cientista_dados/master/data/08.Dados.zip
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 151.101.0.133, 151.101.64.133, 151.101.128.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|151.101.0.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 3150 (3.1K) [application/zip]
Saving to: ‘08.Dados.zip’

08.Dados.zip        100%[===================>]   3.08K  --.-KB/s    in 0s      

2019-04-01 23:34:25 (62.1 MB/s) - ‘08.Dados.z

In [2]:
import numpy as np
print('Numpy version: {}'.format(np.__version__))
import pandas as pd
print('Pandas version: {}'.format(pd.__version__))

# import do csv com dados
base = pd.read_csv('iris.csv')

Numpy version: 1.14.6
Pandas version: 0.22.0


In [3]:
# shape do dataframe
base.shape

# exibe o head
base.head()

sepal length  sepal width  petal length  petal width        class
0           5.1          3.5           1.4          0.2  Iris-setosa
1           4.9          3.0           1.4          0.2  Iris-setosa
2           4.7          3.2           1.3          0.2  Iris-setosa
3           4.6          3.1           1.5          0.2  Iris-setosa
4           5.0          3.6           1.4          0.2  Iris-setosa

In [4]:

# 16. Python: Amostragem Parte I

# seed para o gerador de numeros aleatorios,
# sempre a execucao dara o mesmo resultado
np.random.seed(2345)

# separar amostras
amostra = np.random.choice(a=[0,1], size=150, replace=True, p=[0.5,0.5])

amostra_length = len(amostra)
amostra_one = len(amostra[amostra==1])
amostra_zero = len(amostra[amostra==0])

print('Amostra length: {}'.format(amostra_length))
print('Amostra ones: {}'.format(amostra_one))
print('Amostra zeros: {}'.format(amostra_zero))


Amostra length: 150
Amostra ones: 80
Amostra zeros: 70


In [5]:
# 17. Python: Amostragem Parte II

import pandas as pd
from sklearn.model_selection import train_test_split

iris = pd.read_csv('iris.csv')

# exibe o head
iris.head()

sepal length  sepal width  petal length  petal width        class
0           5.1          3.5           1.4          0.2  Iris-setosa
1           4.9          3.0           1.4          0.2  Iris-setosa
2           4.7          3.2           1.3          0.2  Iris-setosa
3           4.6          3.1           1.5          0.2  Iris-setosa
4           5.0          3.6           1.4          0.2  Iris-setosa

In [6]:
# conta as classes
iris['class'].value_counts()

Iris-versicolor    50
Iris-setosa        50
Iris-virginica     50
Name: class, dtype: int64

In [7]:
# split por estratificação
# dados: iris.iloc[:,0:4]
# atributo que queremos usar para fazer a divisão: iris.iloc[:,4]
# proporcao dos dados usados para o teste: test_size = test_size
# labels para a estatificacao: stratify = iris.iloc[:,4]) 
x, _, y, _ = train_test_split(iris.iloc[:,0:4], 
                     iris.iloc[:,4],
                     test_size = 0.5,
                     stratify = iris.iloc[:,4]) 

print('Quantidade de dados para o teste Iris:\n{}'.format(y.value_counts()))


# show help
# ?train_test_split

Quantidade de dados para o teste Iris:
Iris-versicolor    25
Iris-virginica     25
Iris-setosa        25
Name: class, dtype: int64


In [0]:
infert = pd.read_csv('infert.csv')
infert_value_counts = infert['education'].value_counts()

In [9]:
total_infert = len(infert)
print('total de infert={}'.format(total_infert))


infert_value_counts

education_classes = infert['education'].unique()

for i in education_classes:
  print('{}={}, {:.2f}%'.format(i, infert_value_counts[i], infert_value_counts[i]/total_infert*100))


total de infert=248
0-5yrs=12, 4.84%
6-11yrs=120, 48.39%
12+ yrs=116, 46.77%


In [10]:
# divisao da base de dados
x1, _, y1, _ = train_test_split(infert.iloc[:, 2:9], infert.iloc[:, 1],
                               test_size = 0.6, stratify = infert.iloc[:, 1])


y1.value_counts()

6-11yrs    48
12+ yrs    46
0-5yrs      5
Name: education, dtype: int64

In [11]:
# 18. Python: Amostragem Parte III
# Obter Amostra utilizando a técnica da Amostragem Sistemática

from math import ceil

populacao = 150
amostra = 15

k = ceil(populacao / amostra)

r = np.random.randint(low = 1, high = k + 1, size=1)

print('r={}'.format(r))

acumulador = r[0]
sorteados = []
for i in range(amostra):
  sorteados.append(acumulador)
  acumulador += k
  

  

r=[1]


In [12]:
print('acumulador={}'.format(acumulador))
print('sorteados={}'.format(sorteados))

base = pd.read_csv('iris.csv')
base_final = base.loc[sorteados]
print('acumulador={}'.format(base_final))


acumulador=151
sorteados=[1, 11, 21, 31, 41, 51, 61, 71, 81, 91, 101, 111, 121, 131, 141]
acumulador=     sepal length  sepal width  petal length  petal width            class
1             4.9          3.0           1.4          0.2      Iris-setosa
11            4.8          3.4           1.6          0.2      Iris-setosa
21            5.1          3.7           1.5          0.4      Iris-setosa
31            5.4          3.4           1.5          0.4      Iris-setosa
41            4.5          2.3           1.3          0.3      Iris-setosa
51            6.4          3.2           4.5          1.5  Iris-versicolor
61            5.9          3.0           4.2          1.5  Iris-versicolor
71            6.1          2.8           4.0          1.3  Iris-versicolor
81            5.5          2.4           3.7          1.0  Iris-versicolor
91            6.1          3.0           4.6          1.4  Iris-versicolor
101           5.8          2.7           5.1          1.9   Iris-virginica

# Média - Medida de Centralidade e Variabilidade


## Centralidade

Centralidade:
- Média
- Moda
- Mediana


### Média

Média é a soma de todos os valores e dividido por o número da polulçação

u = Sum x / N

### Moda

Moda é o valor mais frequênte

### Mediana

Mediana é o valor do meio. 

Ordena-se os valores em ordem crescente.
1.  se a quantidade for par, a mediana será a média dos valores nas posições n/2 e n/2+1.
2. se a quantidade for impar, a mediana será o valor na posição (n+1)/2


### Media vs Mediana

Dados = 10,20,30,40,10000

Média: 2020

Mediana: 30


## Variabilidade

Mostra a distância entre os dados.


Variabilidade:
- Variância
- Desvio Padrão
- Amplitude
- Não Centrais: Quartis

### Variância

Mostra a regularidade de um conjunto de dados em relação a média.

Formas diferentes:
- População
- Amostra

### Desvio Padrão

A que distância os dados estão da média.

Raiz quadrada da variância.


### Amplitude

Amplitude é a diferença do maior e do menor valor dos dados.

### Quartis

5, 6, 7, 8, 100

Quartis consegue detectar onde está a maior variação dos dados.

Quartis:
- Q1: 25% dos menores valores
- Q2: 50% igual a mediana
- Q3: 75% dos menores valores
 
## Resumo
 
Dados = 40,18,12,250,30,140,300,40,800
 
Média: 181.111

Q1: 30
 
Q2: 40
 
Q3: 250
 
Desvio Padrão: 253.3079
 
 
## Amostra <> População
 
 
 
|Estatística|Amostra|População|
|--|--|--|
|Média|X|u|
|Desvio Padrão|S|sig|

 

##Fix Numpy Error

# Erro

`module 'numpy' has no attribute 'quantile'`


You must execute:
```bash
!pip install numpy==1.15.1
```

In [13]:
# install numpy 1.15.1
# get rid of the error AttributeError: module 'numpy' has no attribute 'quantile'
!pip install numpy==1.15.1

    100% |████████████████████████████████| 13.9MB 2.8MB/s 
featuretools 0.4.1 has requirement pandas>=0.23.0, but you'll have pandas 0.22.0 which is incompatible.
datascience 0.10.6 has requirement folium==0.2.1, but you'll have folium 0.8.3 which is incompatible.
albumentations 0.1.12 has requirement imgaug<0.2.7,>=0.2.5, but you'll have imgaug 0.2.8 which is incompatible.
  Found existing installation: numpy 1.14.6
    Uninstalling numpy-1.14.6:
      Successfully uninstalled numpy-1.14.6


In [14]:
# 22. Centralidade e Variabilidade

import numpy as np
from scipy import stats

jogadores = [40000, 18000, 12000, 250000,
             30000, 140000, 300000, 40000, 800000]


media = np.mean(jogadores)
mediana = np.median(jogadores)
quartis = np.quantile(jogadores, [0, 0.25, 0.5, 0.75, 1])
desvio_padrao = np.std(jogadores, ddof=1)

print('media={}'.format(media))
print('mediana={}'.format(mediana))
print('quartis={}'.format(quartis))
print('desvio padrao={}'.format(desvio_padrao))

AttributeError: ignored

In [0]:
# describe statistics about jogadores
stats.describe(jogadores)

## Estatisticas a seu favor

Calcular a média e mediana. A média é mais afetada por *outliers*. 

Na sua empresa de compra e venda de carros a partir de valor de mercado, você pode usar a mediana para comprar os carros e vender com base no valor da média. Isso, no caso da mediana < média.

# Probabilidade I

A probabilidade é um valor real entre 0 e 1, pode ser multiplicado por 100 para obter o valor em percentual.

**P = 1** evento certo

**P = 0** evento impossível

## Conceitos

**Experimento**: o que está sendo estudado

**Estado Amostral**: todas as possibilidades de ocorrência do evento

**Evento**: resultados ocorridos

Exemplo:
- Experimento: jogar moeda
- Espaço amostral: cara ou coroa
- Evento: coroa

**Eventos Excludentes**: quando não podem ocorrer ao mesmo tempo

Exemplo: jogar um dado e ser 1 e par

**Eventos não excludentes**: quando podem ocorrer ao mesmo tempo

Exemplo: jogar um dado e ser 2 e par

**Eventos Dependentes**: A ocorrência de um evento afeta o outro. Um tem que ocorrer para depois que o outro ocorra

**Eventos Independentes**: A ocorrência de um evento não afeta o outro.


## Fórmula

**P = Ocorrência esperada / Número de Eventos Possíveis**

Exemplo: jogar uma moeda e dar cara: P = 1/2, P = 0,5 ou 50%

Exemplo: jogar um dado e dar 6: P = 1/6, P = 0,16 ou 16%

Exemplo: jogar um dado e dar 1 ou 6: P = 2/6, P = 0.33 = 33%

Exemplo: jogar um dado e dar impar ou maior que 4 (1,3,5,6); P = 4/6 = 66%

# Probabilidade II

## Eventos Excludentes

Somar as probabilidades dos eventos.

Exemplo: jogar um dado e ser 1 ou par: 1/6 + 3/6 = 4/6 =- 66%

## Eventos Não Excludentes

Somar as probabilidades dos eventos, diminuindo-se as sobreposições.

Exemplo: jogar um dado e ser 2 ou par: 1/6 + 3/6 - 1/6 = 3/6 = 50%


## Eventos Independentes

Mais de um evento: Multiplicação.

Qual a probabilidade de jogar dois dados, e dar 1 e 6 (dois eventos independentes): 1/6 * 1/6 = 1/36 = 0.027 = 2,7%

## Evento Dependentes

Multiplicação da probabilidade de cada evento.

Com 6 cartas na mão, (A, 2 3, 4, 5, 6), qual a probabilidade de primeiro evento tirar A e no segundo evento tirar 4? 1/6 * 1/5 = 0,033 =3,3%

## Probabilidade a longo prazo

JOgando um dado justo 6 vezes, qual a média esperada?
1+2+3+4+5+6/6 = 3,5

A probabilidade é com base no longo prazo

# Distriuição

Usada prinipalmente na teoria da probabilidade

Comportamento de dados aleatórios.

Maior frequência dos dados ao redor da média.

Caracterísiticas de um determinado grupo de dados, por exemplo:
68% dos dados estão a +/- 1 Desvio Padrão da Média
30% dos dados estão a +/- 2 Desvio Padrão da Média

Estes dados com estas caracterísiticas têm x, y e z propriedades que você pode usar para fazer previsões.



# Distribuição Binomial I

Distribuição de Probabilidade Discreta

Permite avaliar a probabilidade de ocorrência de eventos desde que alguns pre requisitos sejam atendidos:
- Número Fixo de Experimentos
- Cada experimento pode ter 2 resultados esperados: **sucesso** ou **fracasso**
- A probabilidade de sucesso deve ser a mesma em cada experimento
- Os experimentos sao independentes

Exemplo: Se eu jogar uma moeda 5 vezes. Qual a probabilidade de dar cara 3 vezes?
- jogar moeda 5 vezes
- cara ou coroa
- P(cara) = P(coroa)
- Eventos independentes

Exemplo: Um cesto tem 10 frutas que pensam entre 3 e 5 quilos. Qual a probabilidade de eu retirar das frutas, uma de 54 quilos e outra de 3 quilos?
- 1 experimento
- duas frutas com 4 e 3 quilos apenas? nao!
- P (4 quilos) != P(4 quillos)? Nao!
- eventos dependentes? Nao!

## Convenções

X = total de sucesso esperado do experimento
p = probabilidade de sucesso
n = número de experimentos
1-p = probabilidades de fracasso

\begin{equation*}
f(x)   = {n \choose x} p^x (1-p)^{ n-x}
\end{equation*}



# Distribuição Binomial II

Exemplo de resolução.

Exemplo: Se eu jogar uma moeda 5 vezes. Qual a probabilidade de dar cara 3 vezes?
- jogar moeda 5 vezes
- cara ou coroa
- P(cara) = P(coroa)
- Eventos independentes

x = 3
p = 0.5
n= 5

\begin{equation*}
f(3)   = {5 \choose 3} p^3 (1-p)^{ 5-3}
\end{equation*}

\begin{equation*}
{5 \choose 3} = \frac{5!}{3!(5-3)!}=\frac{120}{12}=10
\end{equation*}

\begin{equation*}
f(3)   = 10 \cdot (0.125) \cdot (1-0.5)^{ 5-3}
\end{equation*}

\begin{equation*}
f(3)   = 0.3125
\end{equation*}


Exemplo:Se eu passar 4 sinais de quatro tempos cada. Qual a probabilidade de eu pegar 0,12,3 e 4 sinais verdes?
x = 0, 1, 2, 3, 4
p =0.25 (sinais de 4 tempos, 1/4 para cada tempo)
n = 4 (4 sinais)

Todos os semáforos tem a mesma probabilidade.

Resp:
- Nenhum sinal vermelho: 31%
- 1 sinal vermelho: 42%
- 2 sinal vermelho: 21%
- 3 sinal vermelho: 4.6%
- 4 sinal vermelho: 0.39%

Soma = 1 (testamos todas as possibilidades)

Exemplo: Se você fizer a prova de um concurso com 12 questões, chutando todas as questões, qual a probabilidade de acertar 7 questões? (4 alternativas cada questão)

X = 7 acertos
p = 0.25
n = 12

R =0.01147127

É possível recorrer a uma **Tabela de Distribuição Binomial** para identificar o resultado.

![alt text](https://github.com/ronivaldo/udemy-curso-cientista_dados/raw/master/resources/tabela_distribuicao.png "Tabela de Distruibição Binomial")

A probabilidade de ter 7 acertos é a mesma de ter 5 fracassos (cada fracasso a probabilidade é 0.75)

Exemplo: Qual a probabilidade de passar em dois sinais de dois tempos e os dois estarem verdes?

Fazendo manualmente = 1/2 * 1/2 = 0.25

Executando a distribuição binomial no R `dbinomial(2,2,0.5)` = 0.25


In [0]:
# 30. Python: Distribuição Binomial
# distribuição binomial utilizando o python
from scipy.stats import binom

quest = """
Exemplo: Se eu jogar uma moeda 5 vezes. Qual a probabilidade de dar cara 3 vezes?
"""
x = 3
p = 0.5
n = 5
prob = binom.pmf(x, n, p)

print(quest)
print('A probabilidade eh {}'.format(prob))

quest = """
Exemplo:Se eu passar 4 sinais de quatro tempos cada. Qual a probabilidade de eu pegar 0,1,2,3 e 4 sinais verdes?
"""
x = [0, 1, 2, 3, 4]
p = 0.25 #(sinais de 4 tempos, 1/4 para cada tempo)
n = 4 #(4 sinais)

print(quest)
for x1 in x:
  prob = binom.pmf(x1, n, p)
  print('A probabilidade x ={} eh {}'.format(x1, prob))

quest = """
Exemplo: Qual a probabilidade acumulativa de eu passar 4 sinais de quatro tempos cada e pegar 0,1,2,3 ou 4 sinais verdes?
"""
x = 4
p = 0.25
n = 4
prob = binom.cdf(x, n, p)
print(quest)
print('A probabilidade eh {}'.format(prob))

  
quest = """
Exemplo: Qual a probabilidade de passar em dois sinais de dois tempos e os dois estarem verdes?
"""
x = 2
p = 0.5
n = 2
prob = binom.pmf(x, n, p)

print(quest)
print('A probabilidade eh {}'.format(prob))
  
quest = """
Exemplo: Se você fizer a prova de um concurso com 12 questões, chutando todas as questões, qual a probabilidade de acertar 7 questões? (4 alternativas cada questão)
"""
x = 7 # acertos
p = 0.25
n = 12
prob = binom.pmf(x, n, p)

print(quest)
print('A probabilidade eh {}'.format(prob))



# Distribuição Normal I

